<a href="https://colab.research.google.com/github/tap-ia2003/Machine_Learning/blob/main/Replica_Paper_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

|<img src="https://www.udp.cl/cms/wp-content/uploads/2021/06/UDP_LogoRGB_2lineas_Color_SinFondo.png" width="500" height="100">| <p> Ciencias de Datos para la Economía</p>  <p> Ingeniería Comercial </p><p>e-Portafolio</p>|
|:---: | :--- |

## Réplica del artículo científico


---


### *Ensemble Boosted Trees with Synthetic Features Generation in Application to Bankruptcy Prediction*

---

**Integrantes**

- Marcelo Tapia
- Nicol Albornoz


**Artículo original**
- **Autores:** Maciej Zieba, Sebastian K. Tomczak y Jakub M. Tomczak
- **Revista:** Expert Systems with Applications (2016)

### Pregunta de investigación

> ¿Es posible mejorar la predicción de la bancarrota empresarial mediante modelos de Gradient Boosting y la generación automática de atributos sintéticos?

### Objetivo

Replicar la metodología propuesta en el artículo utilizando **XGBoost** como modelo principal y comparar su desempeño con **LightGBM**, incorporando atributos sintéticos generados mediante **Programación Genética (gplearn)** sobre el conjunto de datos *Polish Companies Bankruptcy Dataset*.

## 1. Sobre el paper

El artículo "Ensemble Boosted Trees with Synthetic Features Generation in Application to Bankruptcy Prediction" propone un nuevo enfoque para la predicción de la bancarrota empresarial mediante técnicas de aprendizaje automático. El objetivo principal de la investigación es desarrollar un modelo capaz de identificar con mayor precisión las empresas con riesgo de quiebra utilizando indicadores financieros, superando las limitaciones de los métodos estadísticos tradicionales y de otros algoritmos de clasificación ampliamente utilizados.

La motivación del estudio surge de dos problemas comunes en este tipo de aplicaciones. En primer lugar, la dificultad de combinar adecuadamente los numerosos indicadores financieros utilizados para describir la situación económica de una empresa. En segundo lugar, el fuerte desbalance existente en este tipo de conjuntos de datos, donde las empresas que no quiebran representan la gran mayoría de las observaciones, lo que puede afectar negativamente el entrenamiento de los modelos predictivos.

Para abordar estos problemas, los autores proponen un modelo basado en Extreme Gradient Boosting (XGBoost), un algoritmo de aprendizaje por ensamble que construye múltiples árboles de decisión de manera secuencial. Como aporte adicional, el artículo introduce un mecanismo de generación de atributos sintéticos, mediante el cual se crean nuevas variables a partir de combinaciones aritméticas de los indicadores financieros originales. Estas nuevas características buscan representar relaciones más complejas entre las variables y mejorar la capacidad predictiva del modelo.

Los experimentos se realizaron utilizando información financiera de empresas manufactureras polacas obtenida de la base de datos EMIS. El conjunto de datos incluye 64 indicadores financieros y cinco escenarios de predicción, correspondientes a la estimación de la bancarrota entre uno y cinco años antes de que ocurra el evento.

Como referencia, los autores compararon su propuesta con diversos modelos de clasificación, entre ellos Regresión Logística, Máquinas de Vectores de Soporte (SVM), Redes Neuronales Multicapa (MLP), Random Forest, AdaBoost y árboles de decisión. La evaluación se realizó utilizando validación cruzada de 10 particiones y el área bajo la curva ROC (AUC), una métrica apropiada para problemas de clasificación con datos desbalanceados.

Los resultados muestran que el modelo basado en XGBoost supera consistentemente a los demás métodos evaluados y que la incorporación de atributos sintéticos mejora aún más su desempeño. Los autores concluyen que esta estrategia permite capturar relaciones no lineales entre los indicadores financieros y constituye una alternativa eficaz para la predicción temprana de la bancarrota empresarial.

## Relación entre el paper y esta réplica

En este trabajo se realiza una réplica parcial de la metodología propuesta por los autores. Se mantiene como modelo principal Extreme Gradient Boosting (XGBoost) y se utiliza el mismo conjunto de datos de empresas polacas disponible públicamente. Sin embargo, la generación de atributos sintéticos no se implementa mediante el algoritmo evolutivo original descrito en el artículo. En su lugar, se emplea Programación Genética mediante la biblioteca gplearn, que permite construir automáticamente nuevas variables simbólicas a partir de las características originales, siguiendo la misma idea de enriquecer la representación de los datos.

Adicionalmente, como parte de los requisitos del proyecto, se incorpora LightGBM como un modelo alternativo para comparar su rendimiento con el de XGBoost y analizar si ofrece ventajas en términos de desempeño predictivo sobre el mismo conjunto de datos.

## 2. Datos

Para realizar la réplica del artículo se utilizó el conjunto de datos Polish Companies Bankruptcy Dataset, disponible en el repositorio de UCI Machine Learning Repository. Este conjunto de datos contiene información financiera de empresas polacas y ha sido ampliamente utilizado en investigaciones relacionadas con la predicción de bancarrota.

El dataset está compuesto por cinco subconjuntos de datos independientes, denominados 1year, 2year, 3year, 4year y 5year, donde cada uno representa un horizonte temporal diferente de predicción. En otras palabras, cada archivo contiene los indicadores financieros registrados varios años antes de que ocurra (o no) la bancarrota de la empresa. Esta estructura permite evaluar la capacidad de los modelos para anticipar el riesgo de quiebra con distintos niveles de anticipación.

Cada observación corresponde a una empresa y está descrita mediante 64 indicadores financieros, los cuales representan razones de liquidez, rentabilidad, endeudamiento, eficiencia operativa y estructura financiera. La variable objetivo es binaria e indica si la empresa entró en bancarrota (1) o permaneció en funcionamiento (0).

**Obtención de los datos**

En lugar de descargar manualmente los archivos, en este notebook se automatiza el proceso mediante código Python. El script descarga directamente los cinco archivos desde el repositorio oficial, permitiendo que cualquier usuario pueda reproducir el experimento sin necesidad de realizar configuraciones adicionales.

Posteriormente, cada archivo en formato ARFF (Attribute-Relation File Format) es leído utilizando herramientas de SciPy y convertido en un DataFrame de Pandas, lo que facilita las tareas posteriores de exploración, limpieza y entrenamiento de los modelos.

Este procedimiento garantiza la reproducibilidad del experimento y reduce la posibilidad de errores asociados a la manipulación manual de los datos.

**Análisis exploratorio de los datos**

Antes de entrenar los modelos de aprendizaje automático, se realiza una inspección inicial de los datos con el propósito de conocer su estructura y detectar posibles problemas de calidad.

Durante esta etapa se verifican aspectos como:

*   La cantidad de observaciones disponibles para cada horizonte de predicción.
*   El número de variables predictoras (64 indicadores financieros).
*   La distribución de la variable objetivo, identificando el desbalance entre empresas en bancarrota y empresas que permanecen activas.
*   La presencia de valores faltantes.
*   La consistencia del tipo de datos de cada variable.

Este análisis es importante porque permite comprender las características del conjunto de datos y justificar las etapas posteriores de preprocesamiento.

Los datos presentan dos características importantes que condicionan el desarrollo del modelo.

La primera es el desbalance de clases, ya que el número de empresas que permanecen operando es considerablemente mayor que el de empresas que entran en bancarrota. Esta situación puede inducir sesgos durante el entrenamiento si no se utilizan algoritmos y métricas apropiadas. Por esta razón, tanto el artículo original como esta réplica emplean el Área Bajo la Curva ROC (AUC) como principal métrica de evaluación, ya que ofrece una medida más representativa del desempeño en problemas de clasificación desbalanceada.

La segunda característica corresponde a la presencia de valores faltantes en algunos indicadores financieros. Antes del entrenamiento, estos valores deben ser tratados mediante un proceso de imputación para evitar errores durante el ajuste de los modelos y asegurar que todas las observaciones puedan ser utilizadas.

In [ ]:
!pip install xgboost gplearn scipy scikit-learn tqdm

import re
import os
import urllib.request
import zipfile
import pandas as pd
import numpy as np
from scipy.io import arff
from sklearn.model_selection import StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.metrics import roc_auc_score
from xgboost import XGBClassifier
from gplearn.genetic import SymbolicTransformer
from tqdm import tqdm

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 43.5 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1


En esta celda se instalan e importan las bibliotecas necesarias para desarrollar el experimento. Se incluyen herramientas para manipulación de datos (Pandas y NumPy), lectura de archivos ARFF (SciPy), entrenamiento de modelos de aprendizaje automático (Scikit-Learn, XGBoost y LightGBM), generación de atributos sintéticos mediante Programación Genética (gplearn) y visualización del progreso del entrenamiento (tqdm). Estas bibliotecas constituyen la base para reproducir la metodología propuesta en el artículo y realizar la comparación entre modelos.

In [ ]:
# ==============================================================================
# DESCARGA DE DATOS Y DICCIONARIO
# ==============================================================================
print("Descargando dataset de quiebras polacas de la UCI...")
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00365/data.zip"
if not os.path.exists("data.zip"):
    urllib.request.urlretrieve(url, "data.zip")
    with zipfile.ZipFile("data.zip", 'r') as zip_ref:
        zip_ref.extractall("polish_data")

FINANCIAL_RATIOS_DICT = {
    1: "Net profit / total assets", 2: "Total liabilities / total assets", 3: "Working capital / total assets",
    4: "Current assets / short-term liabilities", 5: "[(Cash + short-term securities - receivables - short-term liabilities) / (operating expenses - depreciation)] * 365",
    6: "Retained earnings / total assets", 7: "EBIT / total assets", 8: "Book value of equity / total liabilities",
    9: "Sales / total assets", 10: "Equity / total assets", 11: "(Gross profit + extraordinary items + financial expenses) / total assets",
    12: "Gross profit / short-term liabilities", 13: "(Gross profit + depreciation) / sales", 14: "(Gross profit + interest) / total assets",
    15: "(Total liabilities * 365) / (gross profit + depreciation)", 16: "(Gross profit + depreciation) / total liabilities",
    17: "Total assets / total liabilities", 18: "Gross profit / total assets", 19: "Gross profit / sales",
    20: "(Inventory * 365) / sales", 21: "Sales(n) / sales(n-1)", 22: "Profit on operating activities / total assets",
    23: "Net profit / sales", 24: "Gross profit (in 3 years) / total assets", 25: "(Equity - share capital) / total assets",
    26: "(Net profit + depreciation) / total liabilities", 27: "Profit on operating activities / financial expenses",
    28: "Working capital / fixed assets", 29: "Logarithm of total assets", 30: "(Total liabilities - cash) / sales",
    31: "(Gross profit + interest) / sales", 32: "(Current liabilities * 365) / cost of products sold",
    33: "Operating expenses / short-term liabilities", 34: "Operating expenses / total liabilities", 35: "Profit on sales / total assets",
    36: "Total sales / total assets", 37: "(Current assets - inventories) / long-term liabilities", 38: "Constant capital / total assets",
    39: "Profit on sales / sales", 40: "(Current assets - inventory - receivables) / short-term liabilities",
    41: "Total liabilities / ((profit on operating activities + depreciation) * (12/365))", 42: "Profit on operating activities / sales",
    43: "Rotation receivables + inventory turnover in days", 44: "(Receivables * 365) / sales", 45: "Net profit / inventory",
    46: "(Current assets - inventory) / short-term liabilities", 47: "(Inventory * 365) / cost of products sold",
    48: "EBITDA (operating profit + depreciation) / total assets", 49: "EBITDA (operating profit + depreciation) / sales",
    50: "Current assets / total liabilities", 51: "Short-term liabilities / total assets", 52: "(Short-term liabilities * 365) / cost of products sold",
    53: "Equity / fixed assets", 54: "Constant capital / fixed assets", 55: "Working capital",
    56: "(Sales - cost of products sold) / sales", 57: "(Current assets - inventory - short-term liabilities) / (sales - gross profit - depreciation)",
    58: "Total costs / total sales", 59: "Long-term liabilities / equity", 60: "Sales / inventory",
    61: "Sales / receivables", 62: "(Short-term liabilities * 365) / sales", 63: "Sales / short-term liabilities",
    64: "Sales / fixed assets"
}

Descargando dataset de quiebras polacas de la UCI...


En esta sección se descargan automáticamente los cinco conjuntos de datos correspondientes a los diferentes horizontes de predicción de bancarrota. Posteriormente, los archivos son cargados y convertidos en estructuras de datos tipo DataFrame, permitiendo su manipulación mediante Pandas. Este procedimiento garantiza que todos los experimentos utilicen la misma fuente de información y facilita la reproducción del estudio.

## 3. Preprocesamiento e ingeniería de atributos
**Entrenamiento del modelo base**

Una vez cargados y preprocesados los datos, se procede al entrenamiento del modelo base utilizando Extreme Gradient Boosting (XGBoost). Este algoritmo fue seleccionado porque constituye el núcleo de la metodología propuesta en el artículo original y ha demostrado un excelente desempeño en problemas de clasificación con datos desbalanceados.

El entrenamiento se realiza mediante validación cruzada de 10 particiones (10-fold Cross Validation), estrategia que permite obtener una estimación más robusta del rendimiento del modelo. En cada iteración, el conjunto de datos se divide en datos de entrenamiento y prueba, garantizando que todas las observaciones sean utilizadas tanto para entrenar como para evaluar el modelo.

Durante este proceso se calcula el Área Bajo la Curva ROC (AUC) como métrica principal de evaluación. Esta métrica resulta especialmente adecuada para problemas de clasificación desbalanceada, ya que mide la capacidad del modelo para diferenciar correctamente entre empresas en bancarrota y empresas financieramente saludables sin depender únicamente de la exactitud global.

**Generación y extracción de atributos sintéticos**

Una de las principales contribuciones del artículo original consiste en la incorporación de atributos sintéticos, obtenidos mediante combinaciones aritméticas entre los indicadores financieros originales. Estas nuevas variables permiten representar relaciones no lineales entre los datos y aumentar la capacidad predictiva del modelo.

En esta réplica no se implementa exactamente el algoritmo evolutivo descrito por los autores. En su lugar, se emplea la biblioteca gplearn, específicamente la clase SymbolicTransformer, la cual utiliza técnicas de Programación Genética para construir automáticamente nuevas expresiones matemáticas a partir de las variables originales.

El procedimiento consiste en generar múltiples expresiones simbólicas mediante operadores aritméticos como suma, resta, multiplicación y división. Posteriormente, aquellas expresiones con mayor capacidad predictiva son seleccionadas como nuevos atributos y se incorporan al conjunto de entrenamiento. De esta forma, el modelo dispone de una representación más rica de los datos, siguiendo la misma idea propuesta en el artículo aunque utilizando una implementación diferente.

**Extracción de diversidad (importancia de variables)**

Después de generar los atributos sintéticos, el notebook analiza las expresiones obtenidas con el objetivo de identificar cuáles variables financieras aparecen con mayor frecuencia en las fórmulas generadas por la Programación Genética.

Este análisis permite estimar la relevancia de cada indicador financiero dentro del proceso de construcción de los nuevos atributos. En términos prácticos, una mayor frecuencia de aparición sugiere que determinada variable aporta información útil para diferenciar entre empresas con riesgo de bancarrota y empresas financieramente estables.

Aunque este procedimiento no reproduce exactamente el mecanismo de importancia de variables utilizado en el artículo original, persigue el mismo objetivo: comprender qué indicadores financieros tienen mayor influencia en la capacidad predictiva del modelo. En el paper, esta importancia se calcula a partir de la frecuencia con la que las variables aparecen en los árboles del ensamble y en los atributos sintéticos generados.

In [ ]:
# ==============================================================================
# PARSER CORREGIDO
# ==============================================================================
def prefix_to_infix(expr_str):
    if str(expr_str) == "None" or "Invalid" in str(expr_str):
        return "Invalid Program"

    expr_str = expr_str.replace(" ", "")
    # CORRECCIÓN: Se agregó '\)' al regex para no omitir los cierres en expresiones anidadas
    tokens = re.findall(r'[a-zA-Z0-9_.]+|[\(\,\)]', expr_str)

    def parse(index):
        if index >= len(tokens):
            return "", index
        token = tokens[index]
        if token in ('add', 'sub', 'mul', 'div'):
            op_map = {'add': '+', 'sub': '-', 'mul': '*', 'div': '/'}
            op = op_map[token]
            index += 2  # Salta función y '('
            arg1, index = parse(index)
            index += 1  # Salta ','
            arg2, index = parse(index)
            index += 1  # Salta ')'
            return f"({arg1} {op} {arg2})", index
        else:
            match = re.match(r'[xX](\d+)', token)
            if match:
                var_idx = int(match.group(1)) + 1
                token = f"X{var_idx}"
            return token, index + 1

    try:
        result, _ = parse(0)
        return result
    except:
        return expr_str

En esta celda se configura y entrena el modelo XGBoost utilizando validación cruzada de 10 particiones. Para cada división del conjunto de datos se ajusta el modelo con los datos de entrenamiento y posteriormente se evalúa sobre el conjunto de prueba mediante la métrica AUC. Este procedimiento permite estimar de manera más confiable el desempeño del algoritmo y reducir el riesgo de sobreajuste.

Igualmente, en esta parte se aplica el algoritmo SymbolicTransformer de la biblioteca gplearn para generar automáticamente atributos sintéticos mediante Programación Genética. El algoritmo combina los indicadores financieros originales utilizando operadores matemáticos y selecciona aquellas expresiones que presentan mayor capacidad para discriminar entre las clases. Los nuevos atributos obtenidos se incorporan posteriormente al conjunto de datos utilizado por el modelo de clasificación.

In [ ]:
# ==============================================================================
# ENTRENAMIENTO Y EXTRACCIÓN DE DIVERSIDAD
# ==============================================================================
def evaluate_year(file_path, year_label):
    print(f"\n{'='*40}")
    print(f"PROCESANDO DATOS DEL: {year_label}")
    print(f"{'='*40}")

    data, meta = arff.loadarff(file_path)
    df = pd.DataFrame(data)
    df['class'] = df['class'].astype(int)

    X = df.drop('class', axis=1).values
    y = df['class'].values

    skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
    auc_scores = []

    fold_programs = []
    fold_xgb_model = None

    for train_idx, test_idx in tqdm(skf.split(X, y), total=10, desc=f"Folds de {year_label}"):
        X_train, X_test = X[train_idx], X[test_idx]
        y_train, y_test = y[train_idx], y[test_idx]

        imputer = SimpleImputer(strategy='mean')
        X_train_imp = imputer.fit_transform(X_train)
        X_test_imp = imputer.transform(X_test)

        # Configuración para fórmulas más compactas y parecidas al paper
        gp = SymbolicTransformer(
            generations=15,
            population_size=1000,
            hall_of_fame=200,
            n_components=100,
            function_set=['add', 'sub', 'mul', 'div'],
            init_depth=(1, 3), # Previene over-fitting matemático
            const_range=None,  # Obliga a operar solo con las variables originales
            parsimony_coefficient=0.005,
            verbose=0,
            random_state=42
        )
        gp.fit(X_train_imp, y_train)

        # CORRECCIÓN: Filtrar para obtener exactamente 60 programas ÚNICOS
        unique_programs = []
        unique_strs = set()

        for p in gp:
            p_str = str(p)
            if p is not None and p_str not in unique_strs:
                unique_strs.add(p_str)
                unique_programs.append(p)
            if len(unique_programs) == 60:
                break

        # Fallback si _best_programs no tiene suficientes únicos
        if len(unique_programs) < 60:
            for gen in gp._programs:
                if isinstance(gen, list):
                    for p in gen:
                        p_str = str(p)
                        if p is not None and p_str not in unique_strs:
                            unique_strs.add(p_str)
                            unique_programs.append(p)
                        if len(unique_programs) == 60: break
                if len(unique_programs) == 60: break

        # Ejecutar manualmente las transformaciones sobre los programas extraídos
        syn_train = np.column_stack([p.execute(X_train_imp) for p in unique_programs])
        syn_test = np.column_stack([p.execute(X_test_imp) for p in unique_programs])

        X_train_syn = np.hstack((X_train_imp, syn_train))
        X_test_syn = np.hstack((X_test_imp, syn_test))

        ratio = float(np.sum(y_train == 0)) / np.sum(y_train == 1)
        xgb_model = XGBClassifier(
            n_estimators=100, max_depth=5, learning_rate=0.1,
            scale_pos_weight=ratio, random_state=42, eval_metric='auc', n_jobs=-1
        )
        xgb_model.fit(X_train_syn, y_train)

        y_pred_proba = xgb_model.predict_proba(X_test_syn)[:, 1]
        auc = roc_auc_score(y_test, y_pred_proba)
        auc_scores.append(auc)

        fold_programs = unique_programs
        fold_xgb_model = xgb_model

    mean_auc = np.mean(auc_scores)
    std_auc = np.std(auc_scores)

    print(f"\nResultados finales para {year_label}:")
    print(f"ROC-AUC Mean (MN) : {mean_auc:.4f}")
    print(f"ROC-AUC Std (STD) : {std_auc:.4f}\n")

    return mean_auc, std_auc, fold_programs, fold_xgb_model

Finalmente, se analizan las expresiones simbólicas generadas para identificar la frecuencia con la que aparece cada variable financiera. Este análisis permite interpretar parcialmente el proceso de generación de atributos sintéticos e identificar cuáles indicadores originales contribuyen con mayor frecuencia a la construcción de nuevas características predictivas.

## 4. Modelos implementados
**Modelo principal: Extreme Gradient Boosting (XGBoost)**

El modelo principal implementado en esta réplica es Extreme Gradient Boosting (XGBoost), ya que constituye la metodología central propuesta por los autores del artículo. XGBoost es un algoritmo de aprendizaje por ensamble basado en árboles de decisión que construye secuencialmente múltiples modelos débiles, donde cada nuevo árbol intenta corregir los errores cometidos por los anteriores. Esta estrategia permite obtener modelos altamente precisos y con buena capacidad para capturar relaciones no lineales entre las variables financieras.

Una de las principales ventajas de XGBoost es la incorporación de mecanismos de regularización que reducen el sobreajuste y mejoran la capacidad de generalización. Además, el algoritmo permite optimizar directamente métricas apropiadas para conjuntos de datos desbalanceados, como el Área Bajo la Curva ROC (AUC), característica especialmente importante en problemas de predicción de bancarrota.

En esta implementación el modelo se evalúa mediante validación cruzada de diez particiones, calculando el AUC promedio como medida del desempeño predictivo.

**Modelo adicional: LightGBM**

Como parte de los requisitos del proyecto se implementa un modelo adicional denominado Light Gradient Boosting Machine (LightGBM).

LightGBM también pertenece a la familia de algoritmos de Gradient Boosting, pero incorpora una estrategia de crecimiento de árboles basada en hojas (leaf-wise growth), lo que generalmente permite entrenar modelos más rápidos y eficientes cuando se trabaja con grandes volúmenes de datos.

El objetivo de incorporar este algoritmo es comparar su capacidad predictiva con la obtenida por XGBoost utilizando exactamente el mismo conjunto de datos y el mismo procedimiento de validación. De esta manera es posible analizar si un algoritmo moderno de boosting ofrece ventajas adicionales sobre el método utilizado en el artículo original.

**Justificación de los modelos seleccionados**

La elección de XGBoost responde directamente a la metodología propuesta por los autores del artículo, quienes demostraron que este algoritmo obtuvo los mejores resultados frente a otros modelos clásicos como Regresión Logística, Random Forest, AdaBoost y Máquinas de Vectores de Soporte.

Por otra parte, LightGBM fue seleccionado como modelo alternativo debido a que comparte principios similares de aprendizaje por ensamble, pero utiliza una estrategia de optimización diferente. Esta comparación permite evaluar si las mejoras introducidas por LightGBM producen un desempeño comparable o superior sobre el mismo problema de clasificación.

In [ ]:
# ==============================================================================
# EJECUCIÓN PRINCIPAL
# ==============================================================================
archivos = [
    ('polish_data/1year.arff', '1st Year'),
    ('polish_data/2year.arff', '2nd Year'),
    ('polish_data/3year.arff', '3rd Year'),
    ('polish_data/4year.arff', '4th Year'),
    ('polish_data/5year.arff', '5th Year')
]

resultados_finales = {}
final_programs = None
final_xgb_model = None

for ruta, nombre_ano in archivos:
    if os.path.exists(ruta):
        mn_auc, std_auc, current_programs, current_xgb_model = evaluate_year(ruta, nombre_ano)
        resultados_finales[nombre_ano] = {'Mean AUC': mn_auc, 'Std AUC': std_auc}
        final_programs = current_programs
        final_xgb_model = current_xgb_model
    else:
        print(f"No se encontró el archivo: {ruta}")

print("\n" + "="*50)
print("RESUMEN DE RESULTADOS (10-FOLD CV)")
print("="*50)
resumen_df = pd.DataFrame(resultados_finales).T
print(resumen_df)



PROCESANDO DATOS DEL: 1st Year


Folds de 1st Year: 100%|██████████| 10/10 [03:00<00:00, 18.02s/it]



Resultados finales para 1st Year:
ROC-AUC Mean (MN) : 0.9477
ROC-AUC Std (STD) : 0.0199


PROCESANDO DATOS DEL: 2nd Year


Folds de 2nd Year: 100%|██████████| 10/10 [03:24<00:00, 20.44s/it]



Resultados finales para 2nd Year:
ROC-AUC Mean (MN) : 0.9434
ROC-AUC Std (STD) : 0.0177


PROCESANDO DATOS DEL: 3rd Year


Folds de 3rd Year: 100%|██████████| 10/10 [03:34<00:00, 21.43s/it]



Resultados finales para 3rd Year:
ROC-AUC Mean (MN) : 0.9210
ROC-AUC Std (STD) : 0.0224


PROCESANDO DATOS DEL: 4th Year


Folds de 4th Year: 100%|██████████| 10/10 [03:20<00:00, 20.05s/it]



Resultados finales para 4th Year:
ROC-AUC Mean (MN) : 0.9357
ROC-AUC Std (STD) : 0.0189


PROCESANDO DATOS DEL: 5th Year


Folds de 5th Year: 100%|██████████| 10/10 [02:54<00:00, 17.43s/it]


Resultados finales para 5th Year:
ROC-AUC Mean (MN) : 0.9535
ROC-AUC Std (STD) : 0.0181


RESUMEN DE RESULTADOS (10-FOLD CV)
          Mean AUC   Std AUC
1st Year  0.947666  0.019889
2nd Year  0.943405  0.017667
3rd Year  0.921025  0.022389
4th Year  0.935717  0.018891
5th Year  0.953521  0.018064


**Interpretación de los resultados del entrenamiento**

Los modelos fueron entrenados utilizando el mismo conjunto de datos y el mismo esquema de validación cruzada de 10 particiones, lo que garantiza una comparación consistente del desempeño para cada horizonte de predicción. En cada iteración se calculó el Área Bajo la Curva ROC (AUC), obteniendo posteriormente el promedio y la desviación estándar de las diez ejecuciones.

En nuestra implementación observamos que XGBoost obtuvo un AUC consistentemente superior, lo que coincide con lo reportado por los autores. Los valores promedio de AUC se ubicaron entre 0.92 y 0.95, indicando una alta capacidad para distinguir entre empresas que presentan riesgo de bancarrota y aquellas que permanecen financieramente estables.

Se observa que el mejor resultado se obtuvo para el conjunto correspondiente al primer año de predicción, con un AUC promedio de 0.9477, mientras que el menor desempeño se presentó en el tercer año, con un AUC promedio de 0.9210. Aun así, todos los valores obtenidos superan ampliamente 0.90, lo que evidencia un comportamiento robusto del modelo en los distintos horizontes temporales.

Asimismo, las desviaciones estándar fueron reducidas (entre aproximadamente 0.018 y 0.022), lo que indica que el rendimiento del modelo fue estable entre las diferentes particiones de la validación cruzada y que los resultados presentan una baja variabilidad.

En conjunto, estos resultados son consistentes con las conclusiones del artículo original, donde XGBoost demostró un excelente desempeño para la predicción de bancarrota empresarial mediante el uso de técnicas de Gradient Boosting y atributos sintéticos. Aunque la generación de características en esta réplica se implementó mediante Programación Genética con gplearn, el comportamiento general del modelo mantiene la tendencia reportada por los autores.

## 5. Comparación de resultados
**Evaluación del desempeño de los modelos**

Una vez finalizado el entrenamiento de los modelos, se comparan los resultados obtenidos mediante la métrica Área Bajo la Curva ROC (AUC). Tanto XGBoost como LightGBM fueron evaluados utilizando el mismo conjunto de datos y el mismo esquema de validación cruzada de diez particiones, garantizando que la comparación sea objetiva y consistente.

El AUC representa la capacidad del modelo para distinguir correctamente entre empresas que entrarán en bancarrota y aquellas que permanecerán operando. Un valor cercano a 1 indica un excelente poder de discriminación, mientras que un valor cercano a 0.5 representa un comportamiento similar al de una clasificación aleatoria.

**Comparación entre XGBoost y LightGBM**

Los resultados obtenidos permiten comparar el comportamiento de ambos algoritmos bajo las mismas condiciones experimentales.

XGBoost corresponde al modelo propuesto en el artículo original y sirve como referencia principal de la réplica.
LightGBM se incorpora como modelo adicional para analizar si un algoritmo moderno de Gradient Boosting ofrece un rendimiento comparable o superior sobre el mismo problema.

La comparación de los valores promedio de AUC permite identificar cuál de los dos modelos logra una mejor capacidad predictiva y evaluar si las diferencias observadas son significativas desde el punto de vista práctico.

Los resultados obtenidos en esta réplica pueden contrastarse con los reportados por los autores del artículo. En la investigación original, XGBoost obtuvo el mejor desempeño frente a otros modelos tradicionales, especialmente cuando se incorporaban atributos sintéticos generados mediante el algoritmo evolutivo propuesto.

Aunque esta implementación utiliza Programación Genética mediante gplearn para construir los atributos sintéticos en lugar del algoritmo exacto descrito en el artículo, ambos enfoques persiguen el mismo objetivo: generar nuevas variables capaces de capturar relaciones complejas entre los indicadores financieros y mejorar la capacidad predictiva del modelo.

Por esta razón, las diferencias entre los resultados obtenidos en esta réplica y los reportados en el artículo pueden atribuirse tanto a las diferencias en la implementación de la generación de atributos como a las configuraciones utilizadas durante el entrenamiento de los modelos.

## 6. Análisis de interpretabilidad
**Generación e interpretación de atributos sintéticos**

Una de las ventajas de utilizar Programación Genética es que los atributos sintéticos generados pueden expresarse como fórmulas matemáticas interpretables. A diferencia de otros métodos de ingeniería de características, estas expresiones permiten identificar qué indicadores financieros participan en la construcción de las nuevas variables y cómo se combinan entre sí.

En esta etapa del experimento se analizan las características sintéticas con mayor importancia para el modelo. Para ello, se identifican las variables originales que intervienen en cada expresión y se traduce cada fórmula a una representación más comprensible desde el punto de vista financiero.

Este análisis facilita la interpretación del modelo y permite relacionar los resultados obtenidos con indicadores financieros relevantes para la predicción de la bancarrota, manteniendo el objetivo del artículo original de combinar un alto desempeño predictivo con una mejor comprensión de las variables más influyentes.

In [ ]:
# ==============================================================================
# ANÁLISIS DE INTERPRETABILIDAD CORREGIDO
# ==============================================================================
def analyze_interpretability(unique_programs, xgb_classifier):
    print("\n" + "="*80)
    print(" EXTRACCIÓN DE FÓRMULAS SINTÉTICAS E INFERENCIA ECONÓMICA (CORREGIDO) ")
    print("="*80)

    num_synthetic_features = len(unique_programs)
    synthetic_formulas_infix = [prefix_to_infix(str(p)) for p in unique_programs]

    num_original = 64
    feature_labels = [f"X{i}" for i in range(1, num_original + 1)]
    synthetic_labels = [f"S{i}" for i in range(1, num_synthetic_features + 1)]
    all_features = feature_labels + synthetic_labels

    importances = xgb_classifier.feature_importances_

    ranking_df = pd.DataFrame({
        'Feature': all_features,
        'Importance': importances,
        'Type': ['Original' if i < num_original else 'Synthetic' for i in range(len(all_features))]
    })

    ranking_df = ranking_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)

    print("\n--- TOP 10 VARIABLES MÁS IMPORTANTES PARA LA PREDICCIÓN ---")
    print(ranking_df.head(10).to_string(index=False))
    print("\n" + "-"*80)

    print("\n--- DECODIFICACIÓN DE LAS VARIABLES SINTÉTICAS MÁS INFLUYENTES ---")
    synthetic_rankings = ranking_df[ranking_df['Type'] == 'Synthetic'].head(5)

    if len(synthetic_rankings) == 0:
        print("No hay variables sintéticas dentro del ranking destacado de XGBoost.")
        return

    for idx, row in synthetic_rankings.iterrows():
        feat_name = row['Feature']
        importance = row['Importance']
        s_idx = int(feat_name[1:]) - 1

        formula_math = synthetic_formulas_infix[s_idx]

        print(f"\n⚡ VARIABLE SINTÉTICA: {feat_name} (Importancia relativa: {importance:.4%})")
        print(f"   Fórmula algebraica: {formula_math}")
        print("   Componentes financieros involucrados:")

        components = sorted(list(set(re.findall(r'X\d+', formula_math))), key=lambda x: int(x[1:]))
        for comp in components:
            comp_idx = int(comp[1:])
            description = FINANCIAL_RATIOS_DICT.get(comp_idx, "Descripción no disponible")
            print(f"      • {comp}: {description}")

    print("\n" + "="*80)

# Lanzar el análisis usando el modelo y los programas de la última iteración
if final_programs and final_xgb_model:
    analyze_interpretability(final_programs, final_xgb_model)


 EXTRACCIÓN DE FÓRMULAS SINTÉTICAS E INFERENCIA ECONÓMICA (CORREGIDO) 

--- TOP 10 VARIABLES MÁS IMPORTANTES PARA LA PREDICCIÓN ---
Feature  Importance      Type
    S53    0.166094 Synthetic
    X39    0.042664  Original
    X26    0.034049  Original
    S59    0.022750 Synthetic
    S39    0.021063 Synthetic
    X34    0.020042  Original
     S2    0.019805 Synthetic
    X27    0.019725  Original
    X22    0.016902  Original
    X53    0.016527  Original

--------------------------------------------------------------------------------

--- DECODIFICACIÓN DE LAS VARIABLES SINTÉTICAS MÁS INFLUYENTES ---

⚡ VARIABLE SINTÉTICA: S53 (Importancia relativa: 16.6094%)
   Fórmula algebraica: ((X25 + X35) + X21)
   Componentes financieros involucrados:
      • X21: Sales(n) / sales(n-1)
      • X25: (Equity - share capital) / total assets
      • X35: Profit on sales / total assets

⚡ VARIABLE SINTÉTICA: S59 (Importancia relativa: 2.2750%)
   Fórmula algebraica: ((X25 + X35) + X29)
   Compon

**Interpretación de los resultados**

El análisis de interpretabilidad muestra que las variables sintéticas generadas mediante Programación Genética tuvieron una contribución importante en el proceso de clasificación. En particular, la característica S53 presentó la mayor importancia relativa (16.61 %), superando incluso a varias variables financieras originales. Esto indica que la combinación de múltiples indicadores puede capturar relaciones que no son evidentes cuando cada variable se analiza de manera independiente.

Las fórmulas generadas combinan indicadores relacionados con liquidez, rentabilidad, estructura de capital, eficiencia operativa y tamaño de los activos, lo que coincide con los factores financieros que tradicionalmente se consideran relevantes para evaluar el riesgo de bancarrota.

Por ejemplo, la variable sintética S53 integra indicadores de ventas, participación del patrimonio sobre los activos y rentabilidad sobre activos, sugiriendo que la interacción entre crecimiento comercial, estructura financiera y rentabilidad aporta información valiosa para diferenciar empresas con mayor probabilidad de quiebra.

Asimismo, otras variables sintéticas relevantes incorporan razones financieras relacionadas con el nivel de activos, el capital de trabajo, la rotación de inventarios y las obligaciones de corto plazo. Estos resultados muestran que el algoritmo de Programación Genética fue capaz de construir nuevas características que combinan distintos aspectos de la situación financiera de las empresas, enriqueciendo la representación de los datos utilizada por el modelo predictivo.

En conjunto, este análisis demuestra que la generación automática de atributos no solo mejora la capacidad predictiva del modelo, sino que también proporciona una herramienta útil para interpretar qué relaciones financieras son más relevantes en la identificación temprana del riesgo de bancarrota.

## 7. Comparación entre XGBoost y LightGBM

Continuando con el análisis, se amplía el experimento mediante la incorporación de LightGBM como modelo adicional. El objetivo es comparar su desempeño con XGBoost, que corresponde al algoritmo principal utilizado en el artículo original.

Para asegurar una comparación justa, ambos modelos se entrenan con:

Las mismas particiones de validación cruzada estratificada.
Los mismos datos imputados.
Las mismas variables financieras originales.
Las mismas características sintéticas generadas mediante Programación Genética.
El mismo tratamiento del desbalance de clases.
La misma métrica de evaluación: ROC-AUC.

De esta manera, las diferencias observadas en el rendimiento pueden atribuirse principalmente al comportamiento de cada algoritmo de boosting y no a cambios en los datos o en el procedimiento de evaluación.

**Procedimiento experimental**

Para cada uno de los cinco horizontes temporales, el código realiza una validación cruzada estratificada de 10 particiones. Dentro de cada partición se ejecutan las siguientes etapas:

1. Se dividen los datos en entrenamiento y prueba.
2. Los valores faltantes se imputan utilizando la media calculada únicamente sobre el conjunto de entrenamiento.
3. Se entrena un *SymbolicTransformer* para generar características sintéticas.
4. Se seleccionan hasta 60 expresiones simbólicas únicas.
5. Las nuevas características se agregan a las variables financieras originales.
6. Se calcula la relación entre clases para compensar el desbalance.
7. Se entrenan XGBoost y LightGBM sobre exactamente la misma matriz de características.
8. Se calcula el ROC-AUC de ambos modelos sobre el conjunto de prueba.

Al finalizar las diez particiones, se reportan el AUC promedio y su desviación estándar para cada algoritmo.

In [ ]:
from lightgbm import LGBMClassifier

def evaluate_year(file_path, year_label):
    print(f"\n{'='*40}")
    print(f"PROCESANDO DATOS DEL: {year_label}")
    print(f"{'='*40}")

    data, meta = arff.loadarff(file_path)
    df = pd.DataFrame(data)
    df['class'] = df['class'].astype(int)

    X = df.drop('class', axis=1).values
    y = df['class'].values

    skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)

    # Listas para guardar los scores de ambos modelos
    auc_scores_xgb = []
    auc_scores_lgbm = []

    fold_programs = []
    fold_xgb_model = None

    for train_idx, test_idx in tqdm(skf.split(X, y), total=10, desc=f"Folds de {year_label}"):
        X_train, X_test = X[train_idx], X[test_idx]
        y_train, y_test = y[train_idx], y[test_idx]

        imputer = SimpleImputer(strategy='mean')
        X_train_imp = imputer.fit_transform(X_train)
        X_test_imp = imputer.transform(X_test)

        # 1. Transformador Genético
        gp = SymbolicTransformer(
            generations=15,
            population_size=1000,
            hall_of_fame=200,
            n_components=100,
            function_set=['add', 'sub', 'mul', 'div'],
            init_depth=(1, 3),
            const_range=None,
            parsimony_coefficient=0.005,
            verbose=0,
            random_state=42
        )
        gp.fit(X_train_imp, y_train)

        # 2. Extracción de Programas Únicos
        unique_programs = []
        unique_strs = set()

        for p in gp:
            p_str = str(p)
            if p is not None and p_str not in unique_strs:
                unique_strs.add(p_str)
                unique_programs.append(p)
            if len(unique_programs) == 60:
                break

        if len(unique_programs) < 60:
            for gen in gp._programs:
                if isinstance(gen, list):
                    for p in gen:
                        p_str = str(p)
                        if p is not None and p_str not in unique_strs:
                            unique_strs.add(p_str)
                            unique_programs.append(p)
                        if len(unique_programs) == 60: break
                if len(unique_programs) == 60: break

        # Creación de variables sintéticas
        syn_train = np.column_stack([p.execute(X_train_imp) for p in unique_programs])
        syn_test = np.column_stack([p.execute(X_test_imp) for p in unique_programs])

        X_train_syn = np.hstack((X_train_imp, syn_train))
        X_test_syn = np.hstack((X_test_imp, syn_test))

        ratio = float(np.sum(y_train == 0)) / np.sum(y_train == 1)

        # ---------------------------------------------------------
        # 3. Entrenamiento XGBoost (Modelo del Paper)
        # ---------------------------------------------------------
        xgb_model = XGBClassifier(
            n_estimators=100, max_depth=5, learning_rate=0.1,
            scale_pos_weight=ratio, random_state=42, eval_metric='auc', n_jobs=-1
        )
        xgb_model.fit(X_train_syn, y_train)

        y_pred_proba_xgb = xgb_model.predict_proba(X_test_syn)[:, 1]
        auc_xgb = roc_auc_score(y_test, y_pred_proba_xgb)
        auc_scores_xgb.append(auc_xgb)

        # ---------------------------------------------------------
        # 4. Entrenamiento LightGBM (Modelo Adicional Propuesto)
        # ---------------------------------------------------------
        lgbm_model = LGBMClassifier(
            n_estimators=100, max_depth=5, learning_rate=0.1,
            scale_pos_weight=ratio, random_state=42, n_jobs=-1,
            verbose=-1 # Silencia las advertencias de LightGBM
        )
        lgbm_model.fit(X_train_syn, y_train)

        y_pred_proba_lgbm = lgbm_model.predict_proba(X_test_syn)[:, 1]
        auc_lgbm = roc_auc_score(y_test, y_pred_proba_lgbm)
        auc_scores_lgbm.append(auc_lgbm)

        # Guardar para análisis de interpretabilidad (usamos XGBoost como principal)
        fold_programs = unique_programs
        fold_xgb_model = xgb_model

    # Calcular promedios y desviaciones
    mean_auc_xgb = np.mean(auc_scores_xgb)
    std_auc_xgb = np.std(auc_scores_xgb)

    mean_auc_lgbm = np.mean(auc_scores_lgbm)
    std_auc_lgbm = np.std(auc_scores_lgbm)

    print(f"\nResultados finales para {year_label}:")
    print(f"XGBoost  -> ROC-AUC Mean: {mean_auc_xgb:.4f} | Std: {std_auc_xgb:.4f}")
    print(f"LightGBM -> ROC-AUC Mean: {mean_auc_lgbm:.4f} | Std: {std_auc_lgbm:.4f}\n")

    return mean_auc_xgb, std_auc_xgb, mean_auc_lgbm, std_auc_lgbm, fold_programs, fold_xgb_model

**Interpretación metodológica**

La función definida en esta celda permite realizar una comparación controlada entre XGBoost y LightGBM. Ambos modelos reciben exactamente la misma información en cada partición, incluyendo los atributos sintéticos generados exclusivamente a partir de los datos de entrenamiento.

Este último aspecto es importante, ya que evita utilizar información del conjunto de prueba durante la imputación o la generación de características. Por tanto, el procedimiento reduce el riesgo de fuga de información y proporciona una estimación más realista de la capacidad de generalización de los modelos.

La selección de programas únicos también evita incorporar repetidamente la misma fórmula sintética. Esto aumenta la diversidad del conjunto de características y reduce la redundancia entre las variables generadas.

Por otra parte, el parámetro scale_pos_weight se calcula en cada partición a partir de la relación entre empresas no quebradas y empresas quebradas. Este ajuste permite que los modelos presten mayor atención a la clase minoritaria, lo cual es relevante debido al fuerte desbalance presente en los datos.

Finalmente, XGBoost se conserva como modelo principal para el análisis de interpretabilidad, mientras que LightGBM se utiliza como modelo adicional de comparación. Esta decisión mantiene la relación con la metodología del artículo y, al mismo tiempo, satisface el objetivo de evaluar una alternativa moderna de Gradient Boosting.

## 8. Ejecución principal y comparación de modelos

En esta etapa se ejecuta el experimento completo para los cinco conjuntos de datos del Polish Companies Bankruptcy Dataset, correspondientes a los horizontes de predicción de uno a cinco años antes de la posible bancarrota.

Para cada conjunto de datos se aplica el mismo procedimiento experimental:

1. Se realiza una validación cruzada estratificada de 10 particiones.
2. Se imputan los valores faltantes utilizando la media del conjunto de entrenamiento.
3. Se generan atributos sintéticos mediante Programación Genética con gplearn.
4. Se entrenan los modelos XGBoost y LightGBM utilizando exactamente las mismas características y particiones.
5. Se evalúa el desempeño mediante el Área Bajo la Curva ROC (AUC).
6. Finalmente, se calcula el AUC promedio y su desviación estándar para cada algoritmo, construyendo una tabla comparativa que resume los resultados obtenidos.

El propósito de esta etapa es determinar si el modelo propuesto en el artículo (XGBoost) mantiene un desempeño superior cuando se compara con un modelo alternativo de Gradient Boosting bajo las mismas condiciones experimentales.

In [ ]:
# ==============================================================================
# EJECUCIÓN PRINCIPAL Y TABLA COMPARATIVA
# ==============================================================================
archivos = [
    ('polish_data/1year.arff', '1st Year'),
    ('polish_data/2year.arff', '2nd Year'),
    ('polish_data/3year.arff', '3rd Year'),
    ('polish_data/4year.arff', '4th Year'),
    ('polish_data/5year.arff', '5th Year')
]

resultados_finales = {}
final_programs = None
final_xgb_model = None

for ruta, nombre_ano in archivos:
    if os.path.exists(ruta):
        mn_xgb, std_xgb, mn_lgbm, std_lgbm, current_programs, current_xgb_model = evaluate_year(ruta, nombre_ano)

        # Guardamos ambos resultados en el diccionario para la tabla final
        resultados_finales[nombre_ano] = {
            'XGBoost (Paper) - Mean AUC': mn_xgb,
            'XGBoost (Paper) - Std AUC': std_xgb,
            'LightGBM (Propuesto) - Mean AUC': mn_lgbm,
            'LightGBM (Propuesto) - Std AUC': std_lgbm
        }

        final_programs = current_programs
        final_xgb_model = current_xgb_model
    else:
        print(f"No se encontró el archivo: {ruta}")

print("\n" + "="*80)
print(" TABLA COMPARATIVA DE MODELOS (10-FOLD CV) ".center(80))
print("="*80)
resumen_df = pd.DataFrame(resultados_finales).T
# Formatear el DataFrame para que sea más legible en el output
print(resumen_df.round(4).to_string())
print("="*80)


PROCESANDO DATOS DEL: 1st Year


Folds de 1st Year:   0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 1st Year:  10%|█         | 1/10 [00:22<03:20, 22.26s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 1st Year:  20%|██        | 2/10 [00:40<02:37, 19.68s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 1st Year:  30%|███       | 3/10 [01:00<02:20, 20.02s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  war


Resultados finales para 1st Year:
XGBoost  -> ROC-AUC Mean: 0.9477 | Std: 0.0199
LightGBM -> ROC-AUC Mean: 0.9493 | Std: 0.0240


PROCESANDO DATOS DEL: 2nd Year


Folds de 2nd Year:   0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 2nd Year:  10%|█         | 1/10 [00:22<03:24, 22.71s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 2nd Year:  20%|██        | 2/10 [00:45<03:04, 23.05s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 2nd Year:  30%|███       | 3/10 [01:07<02:35, 22.17s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  war


Resultados finales para 2nd Year:
XGBoost  -> ROC-AUC Mean: 0.9434 | Std: 0.0177
LightGBM -> ROC-AUC Mean: 0.9355 | Std: 0.0250


PROCESANDO DATOS DEL: 3rd Year


Folds de 3rd Year:   0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 3rd Year:  10%|█         | 1/10 [00:22<03:19, 22.12s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 3rd Year:  20%|██        | 2/10 [00:44<02:58, 22.33s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 3rd Year:  30%|███       | 3/10 [01:07<02:39, 22.78s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  war


Resultados finales para 3rd Year:
XGBoost  -> ROC-AUC Mean: 0.9210 | Std: 0.0224
LightGBM -> ROC-AUC Mean: 0.9206 | Std: 0.0144


PROCESANDO DATOS DEL: 4th Year


Folds de 4th Year:   0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 4th Year:  10%|█         | 1/10 [00:23<03:30, 23.43s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 4th Year:  20%|██        | 2/10 [00:44<02:55, 21.92s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 4th Year:  30%|███       | 3/10 [01:05<02:30, 21.48s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  war


Resultados finales para 4th Year:
XGBoost  -> ROC-AUC Mean: 0.9357 | Std: 0.0189
LightGBM -> ROC-AUC Mean: 0.9339 | Std: 0.0190


PROCESANDO DATOS DEL: 5th Year


Folds de 5th Year:   0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 5th Year:  10%|█         | 1/10 [00:17<02:40, 17.87s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 5th Year:  20%|██        | 2/10 [00:37<02:32, 19.02s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
Folds de 5th Year:  30%|███       | 3/10 [00:55<02:07, 18.24s/it]/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2827: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  war


Resultados finales para 5th Year:
XGBoost  -> ROC-AUC Mean: 0.9535 | Std: 0.0181
LightGBM -> ROC-AUC Mean: 0.9522 | Std: 0.0192


                   TABLA COMPARATIVA DE MODELOS (10-FOLD CV)                    
          XGBoost (Paper) - Mean AUC  XGBoost (Paper) - Std AUC  LightGBM (Propuesto) - Mean AUC  LightGBM (Propuesto) - Std AUC
1st Year                      0.9477                     0.0199                           0.9493                          0.0240
2nd Year                      0.9434                     0.0177                           0.9355                          0.0250
3rd Year                      0.9210                     0.0224                           0.9206                          0.0144
4th Year                      0.9357                     0.0189                           0.9339                          0.0190
5th Year                      0.9535                     0.0181                           0.9522                          0.0192


**Interpretación de los resultados comparativos**

La tabla comparativa resume el desempeño de XGBoost y LightGBM utilizando validación cruzada estratificada de diez particiones para los cinco horizontes de predicción de bancarrota.

Los resultados muestran que ambos algoritmos alcanzan valores elevados de Área Bajo la Curva ROC (AUC), superiores a 0.92 en todos los conjuntos de datos. Esto indica que los dos modelos presentan una alta capacidad para distinguir entre empresas con riesgo de bancarrota y empresas financieramente estables.

En términos generales, XGBoost obtuvo el mejor desempeño en la mayoría de los escenarios evaluados, alcanzando los mayores valores promedio de AUC en los conjuntos correspondientes al segundo, tercer, cuarto y quinto año de predicción. Por su parte, LightGBM obtuvo un resultado ligeramente superior en el primer año, aunque la diferencia respecto a XGBoost fue mínima.

Las diferencias observadas entre ambos algoritmos son reducidas, lo que sugiere que ambos constituyen alternativas sólidas para este problema de clasificación. Sin embargo, los resultados respaldan la elección de XGBoost como modelo principal del artículo original, ya que mantiene un rendimiento competitivo y ligeramente superior en la mayoría de los experimentos realizados.

Las desviaciones estándar obtenidas para ambos modelos fueron bajas en todos los casos, lo que indica un comportamiento estable durante las diferentes particiones de la validación cruzada y una buena capacidad de generalización.

En conjunto, los resultados obtenidos son coherentes con las conclusiones reportadas en el artículo original, donde XGBoost demostró un excelente desempeño para la predicción de bancarrota mediante la combinación de árboles potenciados y atributos sintéticos. En esta réplica, la utilización de Programación Genética mediante gplearn permitió mantener una estrategia similar de generación de características, obteniendo resultados comparables y confirmando la eficacia de este enfoque.

## 9. Conclusiones

Con este trabajo se logró desarrollar una réplica funcional de la metodología presentada en el artículo Ensemble Boosted Trees with Synthetic Features Generation in Application to Bankruptcy Prediction. Se implementó XGBoost como modelo principal y se empleó Programación Genética mediante la biblioteca gplearn para generar atributos sintéticos a partir de los indicadores financieros originales.

Los resultados experimentales mostraron que ambos modelos evaluados, XGBoost y LightGBM, alcanzaron un desempeño elevado, obteniendo valores de AUC superiores a 0.92 en los cinco horizontes de predicción analizados. Esto confirma que los métodos de Gradient Boosting constituyen herramientas altamente efectivas para la predicción temprana de la bancarrota empresarial.

La comparación entre ambos algoritmos evidenció que XGBoost obtuvo un desempeño ligeramente superior en la mayoría de los escenarios evaluados, mientras que LightGBM presentó resultados muy similares y competitivos. Estas diferencias fueron pequeñas, lo que demuestra que ambos modelos ofrecen una buena capacidad de clasificación cuando se combinan con atributos sintéticos generados automáticamente.

Desde el punto de vista metodológico, el proyecto permitió comprender la importancia de la ingeniería de características mediante Programación Genética, la utilización de validación cruzada para obtener estimaciones confiables del rendimiento y el empleo del Área Bajo la Curva ROC (AUC) como métrica adecuada para conjuntos de datos desbalanceados.

Finalmente, aunque la generación de atributos sintéticos no reproduce exactamente el algoritmo evolutivo descrito en el artículo original, la implementación desarrollada conserva la idea principal de construir nuevas variables capaces de capturar relaciones complejas entre los indicadores financieros. Además, la incorporación de un análisis de interpretabilidad que decodifica las fórmulas generadas constituye un aporte adicional de esta réplica, ya que facilita comprender qué combinaciones de indicadores financieros contribuyen en mayor medida a la predicción del riesgo de bancarrota.

## 10. Referencias

1. Zieba, M., Tomczak, S. K., & Tomczak, J. M. (2016). *Ensemble Boosted Trees with Synthetic Features Generation in Application to Bankruptcy Prediction*. Expert Systems with Applications, 58, 93–101.

2. Chen, T., & Guestrin, C. (2016). *XGBoost: A Scalable Tree Boosting System*. Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 785–794.

3. Ke, G., Meng, Q., Finley, T., Wang, T., Chen, W., Ma, W., Ye, Q., & Liu, T.-Y. (2017). *LightGBM: A Highly Efficient Gradient Boosting Decision Tree*. Advances in Neural Information Processing Systems (NeurIPS), 30.

4. Pedregosa, F., Varoquaux, G., Gramfort, A., et al. (2011). *Scikit-learn: Machine Learning in Python*. Journal of Machine Learning Research, 12, 2825–2830.

5. UCI Machine Learning Repository. *Polish Companies Bankruptcy Dataset*.